internal/s3/access.go

```
@@ -25,26 +92,34 @@ func (h *ObjectHandlers) loadBucketForAccess(w http.ResponseWriter, r *http.Requ
+// ensureBucketAction authorizes the given action on an existing bucket.
+// objectKey and listPrefix are optional depending on the action.
+func (h *ObjectHandlers) ensureBucketAction(w http.ResponseWriter, r *http.Request, bucketName, action, objectKey, listPrefix string) bool {
+	_, ok := h.loadBucketForAction(w, r, bucketName, action, objectKey, listPrefix)
+	return ok
 }
 
+// ensureBucket is retained for call sites that only need existence + coarse
+// access. Prefer ensureBucketAction with an explicit action when possible.
+// It uses s3:ListBucket with an empty list prefix so owners/admins and
+// whole-bucket list grantees pass; object operations should use their action.
 func (h *ObjectHandlers) ensureBucket(w http.ResponseWriter, r *http.Request, bucketName string) bool {
```

golangci-lint (2.12.2)

[error] func (*ObjectHandlers).ensureBucket is unused

(unused)